source-mongodb/pull.go

```diff
@@ -255,54 +295,123 @@ func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, bindin
-	cursor, err := collection.Find(ctx, bson.D{})
+	if state.BackfillStartedAt.IsZero() {
+		// This means we are starting backfill for the first time
+		state.BackfillStartedAt = time.Now()
+	}
+
+	var opts = options.Find().SetSort(bson.D{{idProperty, SortAscending}})
+	var filter = bson.D{}
+	if state.BackfillLastId.Validate() == nil {
+		var v interface{}
+		if err := state.BackfillLastId.Unmarshal(&v); err != nil {
+			return fmt.Errorf("unmarshalling backfill_last_id: %w", err)
+		}
+		filter = bson.D{{idProperty, bson.D{{"$gt", v}}}}
+	}
+
+	log.WithFields(log.Fields{
+		"collection": res.Collection,
+		"opts": opts,
+		"filter": filter,
+	}).Info("backfilling documents in collection")
+
+	cursor, err := collection.Find(ctx, filter, opts)
 	if err != nil {
-		return nil, fmt.Errorf("could not run find query on collection %s: %w", res.Collection, err)
+		return fmt.Errorf("could not run find query on collection %s: %w", res.Collection, err)
 	}
 
 	defer cursor.Close(ctx)
 
 	var i = 0
 	for cursor.Next(ctx) {
 		i += 1
+
+		// We decode once into a map of RawValues so we can keep the _id as a
+		// RawValue. This allows us to marshal it as an encoded byte array along
+		// with its type code in the checkpoint, and unmarshal it when we want to
+		// use it to resume the backfill, ensuring the type of the _id is consistent
+		// through this process
+		var rawDoc map[string]bson.RawValue
+		if err = cursor.Decode(&rawDoc); err != nil {
+			return fmt.Errorf("decoding document raw in collection %s: %w", res.Collection, err)
+		}
+		state.BackfillLastId = rawDoc[idProperty]
+
 		var doc bson.M
 		if err = cursor.Decode(&doc); err != nil {
-			return nil, fmt.Errorf("decoding document in collection %s: %w", res.Collection, err)
+			return fmt.Errorf("decoding document in collection %s: %w", res.Collection, err)
 		}
 		doc = sanitizeDocument(doc)
 
 		js, err := json.Marshal(doc)
 		if err != nil {
-			return nil, fmt.Errorf("encoding document %v in collection %s as json: %w", doc, res.Collection, err)
+			return fmt.Errorf("encoding document %v in collection %s as json: %w", doc, res.Collection, err)
 		}
 
 		if err = c.Output.Documents(binding, js); err != nil {
-			return nil, fmt.Errorf("output documents failed: %w", err)
+			return fmt.Errorf("output documents failed: %w", err)
 		}
 
 		if i%CHECKPOINT_EVERY == 0 {
-			if err = c.Output.Checkpoint([]byte("{}"), true); err != nil {
-				return nil, fmt.Errorf("output checkpoint failed: %w", err)
+			var checkpoint = captureState{
+				Resources: map[string]resourceState{
+					resourceId(res): *state,
+				},
+			}
+
+			checkpointJson, err := json.Marshal(checkpoint)
+			if err != nil {
+				return fmt.Errorf("encoding checkpoint to json failed: %w", err)
+			}
+
+			if err = c.Output.Checkpoint(checkpointJson, true); err != nil {
+				return fmt.Errorf("output checkpoint failed: %w", err)
+			}
+
+			if diff, err := OplogTimeDifference(ctx, client); err != nil {
```

Comment:
Conceptually here I think we want to know if the backfill is falling behind how fast it needs to be going in order to not lose the oplog event it needs when starting streaming. Is that something we can conclusively check for in `ChangeStream()`? It kind of looks like a `startAt` can be set, but if that `startAt` is earlier than anything in the oplog there won't be an error. So the scenario is that in the time the backfill took to complete, the change events we needed around `startAt` have fallen off.

It looks a little bit like we need to grab an initial resume token from the change stream prior to starting the backfill from some point earlier than the backfill began, and then resume from that token when transitioning into the stream phase. That way if the token is gone we will error and know that there is missing data. Then, we could periodically check to make sure that resume token still exists while the backfill is on-going to be able to fail fast.

Reply:
We can check to see if the time elapsed since start of backfill is approaching `OplogTimeDifference`, in which case we are getting close to losing the oplog records to start a `ChangeStream` since beginning of backfill